### PR TITLE
Initial PR of IdModel-based indexing

### PR DESCRIPTION
**Pull request by naoyam**
This is the first of a series of IdModel-based indexing PRs. The overall algorithm is exactly what I presented at the nvfuser meeting and is based on the shortest path traversal introduced at #2270. This PR only includes minimum logic for basic indexing. However, pointwise and reduction ops with TID/BID parallelization as well as reshape should work.

Notable missing support:

- Broadcast
- Vectorize, Unswitch, Unroll
- Contiguous indexing

Note that the new indexing interface is not plugged into the lowering yet. Currently, it's only used by the new unit tests. In subsequent PRs, there'll be a switch to use the new indexer instead of the existing one.

Overall tracking PR: #2238

### CODE REVIEW

**Comment by naoyam** [comment]
@jacobhinkle Please review this PR. Of course, feel free to ask any questions. As I mentioned above, this PR has lots of missing pieces but should work for indexing pointwise and reduction fusions like those used in the new unit tests. I'll continue preparing a next set of PRs.

CC: @zasdfgbnm

File: csrc/id_model/id_model.h [hunk 1]
```diff
@@ -153,6 +153,15 @@ class IdModel : public PolymorphicBase {
 
   std::string toString() const;
 
+  bool empty() const {
+    return tvs_.empty();
+  }
+
+  Fusion* fusion() const {
```
[the remark this answers is not in the document]
Comment (naoyam) [reply csrc/id_model/id_model.h:160]: Sure. Thanks for the suggestion.

File: csrc/id_model/indexing.cpp [hunk 2]
```diff
@@ -0,0 +1,430 @@
...
+      cur_contig_stride = SimplifyingIrBuilder::mulExpr(
+          allocation_domains.at(dim)->extent(), cur_contig_stride);
+    } else {
+      // Assume that the tensor should always be a Global memory
+      // tensor if it has non-contig allocation domains
+      NVF_ERROR(tv->getMemoryType() == MemoryType::Global);
+      strides[dim] = getStrideOfGlobalMemoryTensor(tv, (int64_t)dim);
+      cur_contig_stride = strides[dim];
+    }
+  }
```
[the remark this answers is not in the document]
Comment (naoyam) [reply csrc/id_model/indexing.cpp:150]: First of all, this is what we currently do and I'm trying to not change anything unless necessary.

Also, `alloc_stride` is per tensor, so hoisting is limited within each tensor, whereas we keep only a single `logical_size` for all mapped domains, so there should be more opportunities for compiler optimizations.

File: csrc/id_model/indexing.cpp [hunk 3]
```diff
@@ -0,0 +1,430 @@
...
+    return;
+  }
+
+  Fusion* fusion = id_model_.fusion();
+
+  for (auto expr : fusion->exprs()) {
+    if (!ir_utils::isTvOp(expr)) {
+      continue;
+    }
+    auto tv_output = ir_utils::getTvOutput(expr);
```
[the remark this answers is not in the document]
Comment (naoyam) [reply csrc/id_model/indexing.cpp:317]: Right, but that reminds me of the SDPA issue.

File: csrc/id_model/indexing.cpp [hunk 4]
```diff
@@ -0,0 +1,430 @@
...
+      getAllocationDomains(tv, id_model_);
+
+  const auto& index_info = computeIndex(expr, allocation_domains);
+  const auto& index_map = index_info.index_map;
+
+  // Linearize the indices with strides.
+  // TODO: Contiguous indexing
+  Val* index = tv->fusion()->zeroVal();
+  for (const auto i : c10::irange(allocation_domains.size())) {
+    // Traverse from innermost to outermost
```
[the remark this answers is not in the document]
Comment (naoyam) [reply csrc/id_model/indexing.cpp:382]: I don't know, but this is also just following what's currently done.

File: csrc/id_model/id_model.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (naoyam) [reply csrc/id_model/id_model.h:160]: I changed the constructor to set `Fusion*` always.

File: csrc/id_model/indexing.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (jacobhinkle) [reply csrc/id_model/indexing.cpp:150]: > Also, `alloc_stride` is per tensor, so hoisting is limited within each tensor, whereas we keep only a single `logical_size` for all mapped domains, so there should be more opportunities for compiler optimizations.

Makes sense! I forgot about extent simplification that happens later in lowering.

**Review by jacobhinkle — APPROVED** [review]
LGTM. Thanks for the explanations.

**Pull request merged**